### PR TITLE
Support SCA zip scan

### PR DESCRIPTION
**Pull request by NimrodGolan**
### Description

> Introduce a new SCA property 'enabledZipScan' flag which overrides the default 'git-url' scan with zip folder scan.

### References

[320 user story](https://dev.azure.com/CxFlow/CxFlow/_workitems/edit/320)

### Testing

> A new SCA integration test was added

### Checklist

- [X] I have added documentation for new/changed functionality in this PR (if applicable).  *If documentation is a Wiki Update, please indicate desired changes within PR MD Comment*
- [X] All active GitHub checks for tests, formatting, and security are passing
- [X] The correct base branch is being used

### CODE REVIEW

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
Logo<br />**Checkmarx SAST - Scan Summary & Details**
### Cx-SAST Summary
Total of 1361 vulnerabilities<br />High&nbsp;**0 High**<br />Medium&nbsp;**0 Medium**<br />Low&nbsp;**1255 Low**<br />Info&nbsp;**106 Info**<br />
### Violation Summary
#### No policy violation found
### Cx-SAST Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---

File: docs/Integration-with-CxSCA.md [hunk 1]
```diff
@@ -171,4 +172,11 @@ In order to open SCA security tickets, set the bug tracker in cxflow app.yml fil
 * local zip scan:
   * -scan --app=MyApp --cx-team="my-team" --cx-project="project" --f="/Users/myProjects/project"
 * get latest scan results:
-  * --project --app=MyApp --cx-team="my-team" --cx-project="project" ([use 'project' command](https://github.com/checkmarx-ltd/cx-flow/blob/develop/src/main/java/com/checkmarx/flow/dto/ScanRequest.java))
\ No newline at end of file
+  * --project --app=MyApp --cx-team="my-team" --cx-project="project" ([use 'project' command](https://github.com/checkmarx-ltd/cx-flow/blob/develop/src/main/java/com/checkmarx/flow/dto/ScanRequest.java))
+
+
+## <a name="zipFolderScan">SCA ZIP folder scan</a>
+In order to change the default Cx-Flow SCA scan behaviour and to perform a SCA ZIP scan, the next configuration property should be added underneath the sca configuration section:
```
Comment (alex-ko-dev) [inline docs/Integration-with-CxSCA.md:179]: The official product name is CxFlow (without a dash), so it's better to use the name without the dash throughout the docs for consistency.

File: src/main/java/com/checkmarx/flow/service/AbstractASTScanner.java [hunk 2]
```diff
@@ -30,12 +36,21 @@
...
+    private final ScaProperties scaProperties;
+    private final CxRepoFileService cxRepoFileService;
 
     @Override
     public ScanResults scan(ScanRequest scanRequest) {
         ScanResults result = null;
         log.info("--------------------- Initiating new {} scan ---------------------", scanType);
         ScanParams sdkScanParams = toSdkScanParams(scanRequest);
+
+        if (scaProperties.isEnabledZipScan()) {
```
Comment (alex-ko-dev) [inline src/main/java/com/checkmarx/flow/service/AbstractASTScanner.java:48]: I see SCA-specific logic inside AbstractASTScanner. Shouldn't this logic be in SCAScanner?

File: src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/ScaCommonSteps.java [hunk 3]
```diff
@@ -46,4 +47,16 @@ protected ScanRequest getBasicScanRequest(String projectName, String repoWithAut
                 .map(String::trim)
                 .collect(Collectors.toList());
     }
+
+    protected long getTotalHighFindings(SCAResults scaResults) {
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/ScaCommonSteps.java:51]: These methods are only used in ScaZipScanSteps, so it's better to move them there.
getTotalHighFindings and getTotalMediumFindings can be combined into a single method that also includes an assertion.

```
private void validateFindingCount(SCAResults scaResults, String severity) {
   long count = ...
    Assert.assertTrue("...", count > 0);
}
```

File: src/test/resources/cucumber/features/integrationTests/sca/scanResultsProcessing.feature [hunk 4]
```diff
@@ -69,3 +69,10 @@ Feature: Cx-Flow SCA Integration permutation tests
     And bug tracker contains open vulnerabilities
     When resolving on of the vulnerabilities
     Then resolved vulnerability status should getting update to closed
+
+  @SCA_Zip_Scan
+  Scenario: Publish SCA results by zip folder
+    Given scanner is SCA
+    And enabledZipScan property is set with true
+    When initiating a new scan
+    Then returned scan high and medium results are bigger than zero
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/integrationTests/sca/scanResultsProcessing.feature:78]: Is it possible to check that the scan results are indeed based on the zip file? (i.e. check that SCA doesn't use a remote repo directly)

File: src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/zip_scan/ScaZipScanSteps.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        ScanRequest scanRequest = getBasicScanRequest(PROJECT_NAME, GIT_REPO_URL);
+
+        ScanResults scanResults = scaScanner.scan(scanRequest);
+        scaResults = Objects.requireNonNull(scanResults).getScaResults();
+    }
+
+    @Then("returned scan high and medium results are bigger than zero")
+    public void validateResults() {
+        Assert.assertTrue("Expected scan total high results to be a positive number", getTotalHighFindings(scaResults) > 0);
+        Assert.assertTrue("Expected scan total medium results to be a positive number", getTotalMediumFindings(scaResults) > 0);
```
Comment (alex-ko-dev) [inline src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/zip_scan/ScaZipScanSteps.java:65]: validateFindingCount can be used here instead, as suggested above.

File: docs/Integration-with-CxSCA.md (hunk 1 above)
Comment (NimrodGolan) [reply docs/Integration-with-CxSCA.md:179]: Right.

File: src/main/java/com/checkmarx/flow/service/AbstractASTScanner.java (hunk 2 above)
Comment (NimrodGolan) [reply src/main/java/com/checkmarx/flow/service/AbstractASTScanner.java:48]: AbstractASTScanner  holds the logic for the scan flow for both ast & sca.
Moreover, the SCAScanner extends the AbstractASTScanner and it's won't be a good practice to access a son class from parent.

File: src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/ScaCommonSteps.java (hunk 3 above)
Comment (NimrodGolan) [reply src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/ScaCommonSteps.java:51]: fixed.

File: src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/zip_scan/ScaZipScanSteps.java (hunk 5 above)
Comment (NimrodGolan) [reply src/test/java/com/checkmarx/flow/cucumber/integration/sca_scanner/zip_scan/ScaZipScanSteps.java:65]: Yes, I changed it.

**Review by alex-ko-dev — APPROVED** [review]

**Pull request merged**